### PR TITLE
FEATURE: Dynamic EditPreview Modes

### PR DESCRIPTION
**Pull request by dfeyer**
This change introduce a new configuration for document node type,
to configure available editing / preview mode per document type. The
configuration of the editing modes stay in `Settings.yaml`, and you
can configure each document node type:

```
'TYPO3.Neos:Document':
  ui:
    editPreviewModes:
      'TYPO3.Neos:inPlace': 
        enabled: TRUE
      'TYPO3.Neos:rawContent':
        enabled: TRUE
      'TYPO3.Neos:desktop':
        enabled: TRUE
```

You can override the position of the mode:

```
'TYPO3.Neos:Document':
  ui:
    editPreviewModes:
      'TYPO3.Neos:rawContent':
        position: 10
```

This change also name space to name of the mode to avoid collision with
modes registered by other packages.

### CODE REVIEW

**Comment by dfeyer** [comment]
http://g.recordit.co/U5QRPM0Jbo.gif

**Comment by dfeyer** [comment]
- [x] Add argument nodeType in SettingsController::editPreviewAction
- [x] Add support to override modes ordering (position)
- [x] Hide editing / preview if no modes are available
- [ ] Improve fallback if the current mode is not available to a target document (CSS issue, ...)

**Comment by kitsunet** [comment]
this shoudl go in the `ui` part of the configuration.

File: TYPO3.Neos/Configuration/Settings.yaml [hunk 1]
```diff
@@ -302,24 +302,24 @@ TYPO3:
       defaultEditPreviewMode: 'inPlace'
       editPreviewModes:
         # Live mode is only configured here for consistency. You shouldn't change it.
-        live:
+        'TYPO3.Neos:live':
```
Comment (aertmann) [inline TYPO3.Neos/Configuration/Settings.yaml]: this is breaking, so it needs a configuration migration

**Comment by aertmann** [comment]
nice one @dfeyer

**Comment by mficzel** [comment]
I like the feature but i doubt that ordering via integer is a good idea. To me that does not look consistent with how neos and flow handles ordering and configuration. 

Is it really necessary to alter the ordering of edit/preview-modes that was defined in settings per document-type? I think that would confuse editors.

**Comment by dfeyer** [comment]
> Is it really necessary to alter the ordering of edit/preview-modes that was defined in settings per document-type? I think that would confuse editors.

@mficzel I dont like to current syntax too, but see this as a useful feature for some cases. I you start to have really specific editing preview modes per node type changing the order can be more easy in the node type configuration because the global settings configuration can be big.

What do you think about:

    'TYPO3.Neos:Document':
      ui:
        editPreviewModes:
          'TYPO3.Neos:inPlace':
            enabled: TRUE
            position: 100
          'TYPO3.Neos:rawContent':
            enabled: TRUE
            position: 200
          'TYPO3.Neos:desktop':
            enabled: TRUE
            position: 300

And what about wildcard support (to all to explicitly disabled all modes and enable 1+ specific modes):

    'TYPO3.Neos:Document':
      ui:
        editPreviewModes:
          '*':
            enabled: FALSE
          'TYPO3.Neos:inPlace':
            enabled: TRUE
            position: 100

File: TYPO3.Neos/Configuration/Settings.yaml (hunk 1 above)
Comment (dfeyer) [inline TYPO3.Neos/Configuration/Settings.yaml]: > # Live mode is only configured here for consistency. You shouldn't change it.

But if we can not change it, why not removing this mode from the settings and apply this mode only in code ? And make the "live" key a reserved key.

Comment (aertmann) [inline TYPO3.Neos/Configuration/Settings.yaml]: actually meant that prefixes all of them are breaking, not just live..

regarding only having it in the code would mean you cannot change the title

**Comment by aertmann** [comment]
I don't see any problem making the sorting flexible and agree it can be useful. Regarding the integer syntax, that's okay since that's already in use multiple times. However it can be improved by doing the sorting server side if possible in an easy way, otherwise I started porting the `PositionalArraySorter` from PHP to JS but haven't finished. If anyone wants they can do that and improve the other places where integer sorting is used currently (https://gist.github.com/aertmann/7686223d3943f428074ef3cb6db3ea21)

Regarding confusing editor, that's really up to the integrator to decide. If it makes sense to do that, it should make sense to the editors, but I don't think it's we can make that call.

Also if the wildcard thing makes sense, go ahead, just make sure you use the same code / port the code from the node types to make sure the ordering is solid.

File: TYPO3.Neos/Configuration/Settings.yaml (hunk 1 above)
Comment (dfeyer) [inline TYPO3.Neos/Configuration/Settings.yaml]: Right, can you point me to an existing configuration migration, not sure how to proceed in clean way.

Comment (aertmann) [inline TYPO3.Neos/Configuration/Settings.yaml]: https://github.com/neos/neos-development-collection/blob/master/TYPO3.Neos.NodeTypes/Migrations/Code/Version20130911165500.php
https://github.com/neos/neos-development-collection/blob/master/TYPO3.Neos/Migrations/Code/Version20141218134700.php
https://github.com/neos/neos-development-collection/blob/master/TYPO3.Neos/Migrations/Code/Version20141113115300.php
https://github.com/neos/neos-development-collection/blob/master/TYPO3.TYPO3CR/Migrations/Code/Version20140911160326.php
https://github.com/neos/neos-development-collection/blob/master/TYPO3.TYPO3CR/Migrations/Code/Version20130523180140.php

**Comment by kitsunet** [comment]
As this hasn't left WIP status by now I remove the Important for 2.3 label as I doubt it will still be ready.

**Comment by dfeyer** [comment]
I remove the WIP status, currently moving from one document to an other where the current editing/preview mode does not exist, reload the backend, that can be improved be require a lots of change and it's out of my current capability. @aertmann If you have an idea how to improve this, your are welcome

I create a PR for the Demo site, to test this feature https://github.com/neos/Neos.Demo/pull/15

**Comment by dfeyer** [comment]
@aertmann Core migration is now include, I think we are ready for last review batch and if some one as an idea to improve the switch between document with different edit/prievew mode, feel free to push something.

**Comment by dfeyer** [comment]
I close this one and will start a new one for the react UI

**Pull request closed without merging**